### DISCUSSION TITLE
Run tests in a module discovery

### DISCUSSION THREAD

**Discussion by brycegbrazen** [pytest-dev/pytest#12243] [Q&A] [answered]
Hi all,

Looking at the "Run tests in a module" section of the [docs](https://docs.pytest.org/en/8.1.x/how-to/usage.html#specifying-which-tests-to-run), it doesn't mention if the path to a module must be:
- An absolute path
- A relative path
- Just a file basename that is recursively found starting with the `testpaths`

Can someone clarify the intended behavior? Perhaps this is something we can add explicit documentation for?

My specific issue is with bullet 3 above, where I'm trying to pass a basename to pytest via `pytest basename.py`, and I get a:
```
ERROR: file or directory not found: basename.py
```

**Comment by nicoddemus** [comment]
It works for absolute or relative paths. If it is not working for you it must be something else, please post a full example.

**Comment by brycegbrazen** [comment]
My folder structure looks like:
```
├── src
│   ├── api
│   │   ├── checks
│   │   │   ├── maya
│   │   │   │   ├── validate_master.py
├── tests
│   ├── __init__.py
│   ├── api
│   │   ├── __init__.py
│   │   ├── checks
│   │   │   ├── __init__.py
│   │   │   ├── maya
│   │   │   │   ├── __init__.py
│   │   │   │   ├── test_validate_master.py
├── pytest.ini
```
The contents of my `pytest.ini` file look like this:
```
[pytest]
testpaths = tests
```

From the root of the folder, I want to run
```
pytest test_validate_master.py
```
where pytest can find a file with that base name under the configured test paths.
However, I get an error:
```
ERROR: file or directory not found: test_validate_master.py
```

If I give pytest a relative path from my current directory (the root folder), it does successfully find the file:
```
pytest tests/api/checks/maya/test_validate_master.py

...
ModuleNotFoundError: No module named 'six'
```

**Reply by nicoddemus** [reply]
(It helps to post this as a reply to my other reply, to keep things in context)

You can run `pytest tests` and pytest will look for test files in that folder, passing the full path to a file also works.

`ModuleNotFoundError: No module named 'six'` means that `six` is not installed in your environment, and is not related ot pytest.

**Reply by brycegbrazen** [reply]
Apologies for not posting as a reply initially.

But just to clarify, is it possible to just pass a file basename if you want to run just a single file? Just curious if it will do a recursive search for the file.

**Reply by The-Compiler** [reply]
No.

**Reply by nicoddemus** [reply, marked as answer]
> But just to clarify, is it possible to just pass a file basename if you want to run just a single file? Just curious if it will do a recursive search for the file.

Not the filename directly (as @The-Compiler commented), but it is possible to use the `-k` flag (`pytest -k basename`) and pytest will collect all files and run only those that contain `basename` in their "node id".

**Reply by brycegbrazen** [reply]
Gotcha. Thanks for the info! One other question, is it possible to use any form of globbing (the * operator) just via the positional argument?

Something like:
`pytest tests/*test_validate_master.py`
or
`pytest **test_validate_master.py`
to search recursively?

**Reply by nicoddemus** [reply]
Not natively.

**Reply by The-Compiler** [reply]
If you're on bash, `pytest **/test_validate_master.py` will work - but that's a bash thing, not a pytest thing.
